core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -283,6 +290,9 @@ private void advanceToNextPage() {
           }
         }
       }
+      if (pageToFree != null) {
+        freePage(pageToFree);
```

Comment:
is it possible that this page is already freed by another consumer?

Reply:
the `MapIterator.spill` will be called by `BytesToBytesMap.spill` which will be called by other consumers.

Reply:
The page to free (`currentPage`) is removed from `dataPages` and advanced to next page when locking on the `MapIterator` object. The locking will prevent the same page to be freed by calling `spill` at another consumer.

Reply:
ah I misread the code. Can you move `synchronized` from `spill` method and put it in method definition?

Reply:
Moved.

Reply:
hmm, is it better to put `freePage` into a `finally` block?

Reply:
I'm not sure if that code branch can throw exception, maybe a safer change is to free the page in `finally`?

Reply:
yea, I'm not sure too. But a `finally` seems to be safer. I can make a follow-up for this.